Breaking Connections

Sploder12 opened the pull request:
Proposed changes
-------------------------------------------
Add "Breaking" connections that can only be used once.
Change Connections to use dedicated Connection classes.

Types of changes
-------------------------------------------
- [X] New feature (non-breaking change which adds functionality)

Checklist
-------------------------------------------
- [X] My changeset covers only what is described above (no extraneous changes)
- [X] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)
- [X] Any dependent changes have been merged downstream

Further comments
-------------------------------------------
Would like connection parsing to use more pattern matching as opposed to equality checks. However, more discussion is needed.

Sploder12 committed:
Added breaking connections

Added connections that can only be used once. Reworked connection API and parser to support this without a massive if-else chain. Some issues need discussion such as the abnormal syntaxes found in crpg.py

Sploder12 committed:
Reverted to Equality Checks

the strange "valid" syntaxes for connections is unwanted. Until we get a better system, equality checks it is

emilkovacev commented on builtin.py:
```diff
@@ -9,19 +9,48 @@ def __init__(self):
+        super().__init__(from_node, to_node)
+        
+    def on_select(self):
+        self.from_node.connections.remove(self)
```
I really like the abstraction of connections, and the versatility it allows for making new types of connections

emilkovacev commented on builtin.py:
```diff
@@ -9,19 +9,48 @@ def __init__(self):
+        pass
+
+    def __str__(self):
+        return str(self.to_node)
```
Perhaps this should include both nodes? Unless you're using this specifically somewhere it would be more descriptive

emilkovacev commented on crpg.py:
```diff
@@ -102,12 +94,22 @@ def generate(filename):
-
-        return game
 
+            if (connection.group(2) == "->"):
+                game.add_connection(node_a, node_b)
+            elif (connection.group(2) == "<->"):
+                game.add_connection(node_a, node_b)
+                game.add_connection(node_b, node_a)
+            elif (connection.group(2) == "\->"):
+                game.add_connection(node_a, node_b, BreakingConnection)
+            elif (connection.group(2) == "<-\->"):
+                game.add_connection(node_a, node_b, BreakingConnection)
+                game.add_connection(node_b, node_a)
+            elif (connection.group(2) == "<-/\->"):
+                game.add_connection(node_a, node_b, BreakingConnection)
+                game.add_connection(node_b, node_a, BreakingConnection)
```
We could create a mapping of strings to functions, and execute the function based on the value of group(2), just so we avoid the if chain

emilkovacev replied:
This way, we could also assert that group(2) is in the mapping keys before moving forward, without having to make the regex too complicated

Sploder12 replied:
I have changed it to use a mapping of str to functions. Also I added exceptions for invalid inputs.
ALSO I changed the regex so that trailing whitespace is ignored, we should set up some way to tell the user where the .dl has issues because something like trailing whitespace is really hard to spot.

emilkovacev replied:
What do you think of `r"^\s*(\w+)\s+([^\s]+)\s+(\w+)\s*$"` as an alteration of the regex? This forces at least one space between the node identifiers, but more importantly allows non-numeric identifiers (kind of like variables).

emilkovacev replied:
The parser seems like it may grow large enough to where it warrants its own file, where we can program in better error messages. I also mentioned in someone else's PR, that it could be cool to use click CLI to run everything, that way the CLI could give appropriate errors based on file parsing

Sploder12 replied:
I like that regex more than the current one mainly due to the non-numeric identifiers.

Definitely needs it's own file. I have not used click CLI so I have no opinions on that. However, I think those changes should be in a separate PR.

emilkovacev replied:
I'd be happy to work on that after we get this through then, or at least make that an issue

emilkovacev reviewed (changes requested):
This opens the door for us to make custom connection types for the future, I'm looking forward to implementing a few myself!

Sploder12 committed:
Changed __str__ for Connections

this should properly represent the connections as strings now.

Sploder12 committed:
Remove connection elif

removed the big chain of conditionals and replaced with a dictionary or lambda functions. Also added error raising and modified the Regex so that trailing whitespace would not cause an exception.

Sploder12 committed:
Removed whitespace

whoops, forgot I left that there

emilkovacev commented:
This is ready to be merged now, the changes add really important functionality onto the API

emilkovacev reviewed (approved):
This is ready to be merged now, the changes add really important functionality onto the API.
